pmtiles/server.go

```diff
@@ -448,48 +485,85 @@ func parseMetadataPath(path string) (bool, string) {
 	return false, ""
 }
 
-// Get a response for the given path.
-// Return status code, HTTP headers, and body.
-func (server *Server) Get(ctx context.Context, path string) (int, map[string]string, []byte) {
-	httpHeaders := make(map[string]string)
+func (server *Server) get(ctx context.Context, path string) (archive, handler string, status int, headers map[string]string, data []byte) {
+	handler = ""
+	archive = ""
+	headers = make(map[string]string)
 	if len(server.cors) > 0 {
-		httpHeaders["Access-Control-Allow-Origin"] = server.cors
+		headers["Access-Control-Allow-Origin"] = server.cors
 	}
 
 	if ok, key, z, x, y, ext := parseTilePath(path); ok {
-		return server.getTile(ctx, httpHeaders, key, z, x, y, ext)
-	}
-	if ok, key := parseTilejsonPath(path); ok {
-		return server.getTileJSON(ctx, httpHeaders, key)
-	}
-	if ok, key := parseMetadataPath(path); ok {
-		return server.getMetadata(ctx, httpHeaders, key)
+		archive, handler = key, "tile"
+		status, headers, data = server.getTile(ctx, headers, key, z, x, y, ext)
+	} else if ok, key := parseTilejsonPath(path); ok {
+		archive, handler = key, "tilejson"
+		status, headers, data = server.getTileJSON(ctx, headers, key)
+	} else if ok, key := parseMetadataPath(path); ok {
+		archive, handler = key, "metadata"
+		status, headers, data = server.getMetadata(ctx, headers, key)
+	} else if path == "/" {
+		handler, status, data = "/", 204, []byte{}
+	} else {
+		handler, status, data = "404", 404, []byte("Path not found")
 	}
 
-	if path == "/" {
-		return 204, httpHeaders, []byte{}
-	}
+	return
+}
 
-	return 404, httpHeaders, []byte("Path not found")
+// Get a response for the given path.
+// Return status code, HTTP headers, and body.
+func (server *Server) Get(ctx context.Context, path string) (int, map[string]string, []byte) {
+	tracker := server.metrics.startRequest()
+	archive, handler, status, headers, data := server.get(ctx, path)
+	tracker.finish(archive, handler, status, len(data), true)
+	return status, headers, data
+}
+
+type loggingResponseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func (lrw *loggingResponseWriter) WriteHeader(code int) {
+	lrw.statusCode = code
+	lrw.ResponseWriter.WriteHeader(code)
 }
 
 // Serve an HTTP response from the archive
 func (server *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) int {
-	statusCode, headers, body := server.Get(r.Context(), r.URL.Path)
+	tracker := server.metrics.startRequest()
+	if r.Method == http.MethodOptions {
+		if len(server.cors) > 0 {
+			w.Header().Set("Access-Control-Allow-Origin", server.cors)
+		}
+		w.WriteHeader(204)
+		tracker.finish("", r.Method, 204, 0, false)
```

Comment:
The "false" argument at the end turns skips logging response size and latency metrics since I figure that's not really interesting if it's always 0 for both.  I could include it though if you think it might be interesting.

Reply:
I envision the `--cors` option being used only for development or non-production since it will always be incomplete relative to Caddy or a CDN. So those metrics won't be interesting.